### PR TITLE
Use lazy covariance in `BatchedMultiOutputGPyTorchModel.posterior`

### PR DESCRIPTION
**Pull request by valtron**
Part of the fix for #971.

## Motivation

This prevents the evaluation of the entire covariance matrix for multi-output posterior evaluation, so `SingleTaskGP.posterior` is O(n) instead of O(n^2).

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

I skimmed it. Agreed to CLA.

## Test Plan

I tested the changes locally with monkey-patches. Here is the script:

```python
import contextlib
from time import perf_counter as time

import numpy as np
import torch
from botorch.models.gp_regression import SingleTaskGP

def main():
	rng = np.random.default_rng(0)
	
	d = 2
	k = 2
	X_train = torch.tensor(rng.uniform(size = (5, d)))
	Y_train = torch.tensor(rng.uniform(size = (5, k)))
	
	gp = SingleTaskGP(X_train, Y_train)
	
	print("{:>5} | {:>6} | {:>6} | {:>7}".format("n", "orig", "patch", "speedup"))
	
	for n_eval in [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096]:
		time = Timer()
		
		X_eval = torch.tensor(rng.uniform(size = (n_eval, d)))
		
		p1 = gp.posterior(X_eval)
		with time('unpatched'):
			for _ in range(10):
				_ = gp.posterior(X_eval)
		
		with monkey_patches():
			p2 = gp.posterior(X_eval)
			with time('patched'):
				for _ in range(10):
					_ = gp.posterior(X_eval)
		
		assert torch.allclose(p1.mean, p2.mean)
		assert torch.allclose(p1.variance, p2.variance)
		
		print("{:>5} | {:>6.4f} | {:>6.4f} | {:>6.1f}x".format(
			n_eval, time['unpatched'] / 10, time['patched'] / 10,
			time['unpatched'] / time['patched'],
		))

class Timer:
	def __init__(self):
		self.times = {}
	
	def __getitem__(self, key):
		return self.times[key]
	
	@contextlib.contextmanager
	def __call__(self, key):
		t0 = time()
		yield
		t1 = time()
		self.times[key] = t1 - t0

@contextlib.contextmanager
def monkey_patches():
	from gpytorch.kernels import Kernel
	from gpytorch.lazy import LazyEvaluatedKernelTensor
	from botorch.models.gpytorch import BatchedMultiOutputGPyTorchModel
	
	orig_Kernel_getitem = Kernel.__getitem__
	orig_LazyEvaluatedKernelTensor_unsqueeze_batch = LazyEvaluatedKernelTensor._unsqueeze_batch
	orig_BatchedMultiOutputGPyTorchModel_posterior = BatchedMultiOutputGPyTorchModel.posterior
	
	try:
		Kernel.__getitem__ = patch_Kernel_getitem
		LazyEvaluatedKernelTensor._unsqueeze_batch = patch_LazyEvaluatedKernelTensor_unsqueeze_batch
		BatchedMultiOutputGPyTorchModel.posterior = patch_BatchedMultiOutputGPyTorchModel_posterior
		yield
	finally:
		BatchedMultiOutputGPyTorchModel.posterior = orig_BatchedMultiOutputGPyTorchModel_posterior
		LazyEvaluatedKernelTensor._unsqueeze_batch = orig_LazyEvaluatedKernelTensor_unsqueeze_batch
		Kernel.__getitem__ = orig_Kernel_getitem

def patch_Kernel_getitem(self, index):
	from copy import deepcopy
	
	if len(self.batch_shape) == 0:
		return self
	new_kernel = deepcopy(self)
	index = index if isinstance(index, tuple) else (index,)
	
	for param_name, param in self._parameters.items():
		new_kernel._parameters[param_name].data = param.__getitem__(index)
		ndim_removed = len(param.shape) - len(new_kernel._parameters[param_name].shape)
		new_batch_shape_len = len(self.batch_shape) - ndim_removed
		obs = new_kernel.batch_shape
		new_kernel.batch_shape = new_kernel._parameters[param_name].shape[:new_batch_shape_len]
	
	for sub_module_name, sub_module in new_kernel.named_sub_kernels():
		new_kernel._modules[sub_module_name] = sub_module.__getitem__(index)
	
	return new_kernel

def patch_LazyEvaluatedKernelTensor_unsqueeze_batch(self, dim):
	return self[(slice(None),) * dim + (None,)]

def patch_BatchedMultiOutputGPyTorchModel_posterior(
	self, X, output_indices = None, observation_noise = False, **kwargs,
):
	from gpytorch.distributions import MultitaskMultivariateNormal, MultivariateNormal
	from gpytorch.likelihoods.gaussian_likelihood import FixedNoiseGaussianLikelihood
	from botorch.models.utils import add_output_dim, gpt_posterior_settings
	from botorch.posteriors.gpytorch import GPyTorchPosterior
	
	self.eval()  # make sure model is in eval mode
	# input transforms are applied at `posterior` in `eval` mode, and at
	# `model.forward()` at the training time
	X = self.transform_inputs(X)
	
	with gpt_posterior_settings():
		# insert a dimension for the output dimension
		if self._num_outputs > 1:
			X, output_dim_idx = add_output_dim(
				X=X, original_batch_shape=self._input_batch_shape
			)
		mvn = self(X)
		if observation_noise is not False:
			if torch.is_tensor(observation_noise):
				# TODO: Validate noise shape
				# make observation_noise `batch_shape x q x n`
				obs_noise = observation_noise.transpose(-1, -2)
				mvn = self.likelihood(mvn, X, noise=obs_noise)
			elif isinstance(self.likelihood, FixedNoiseGaussianLikelihood):
				# Use the mean of the previous noise values (TODO: be smarter here).
				noise = self.likelihood.noise.mean().expand(X.shape[:-1])
				mvn = self.likelihood(mvn, X, noise=noise)
			else:
				mvn = self.likelihood(mvn, X)
		if self._num_outputs > 1:
			mean_x = mvn.mean
			covar_x = mvn.lazy_covariance_matrix
			output_indices = output_indices or range(self._num_outputs)
			mvns = [
				MultivariateNormal(
					mean_x.select(dim=output_dim_idx, index=t),
					covar_x[(slice(None),) * output_dim_idx + (t,)],
				)
				for t in output_indices
			]
			mvn = MultitaskMultivariateNormal.from_independent_mvns(mvns=mvns)
	
	posterior = GPyTorchPosterior(mvn=mvn)
	if hasattr(self, "outcome_transform"):
		posterior = self.outcome_transform.untransform_posterior(posterior)
	return posterior

if __name__ == '__main__':
	main()
```

## Related PRs

In GPyTorch:
- cornellius-gp/gpytorch#1811
- cornellius-gp/gpytorch#1813

### CODE REVIEW

**Comment by valtron** [comment]
Timing:
```
    n |   orig |  patch | speedup
    1 | 0.0012 | 0.0010 |    1.1x
    2 | 0.0012 | 0.0011 |    1.1x
    4 | 0.0020 | 0.0011 |    1.9x
    8 | 0.0012 | 0.0011 |    1.1x
   16 | 0.0012 | 0.0011 |    1.1x
   32 | 0.0013 | 0.0011 |    1.2x
   64 | 0.0015 | 0.0013 |    1.1x
  128 | 0.0019 | 0.0015 |    1.3x
  256 | 0.0064 | 0.0054 |    1.2x
  512 | 0.0273 | 0.0048 |    5.7x
 1024 | 0.1313 | 0.0112 |   11.8x
 2048 | 0.5753 | 0.0125 |   46.1x
 4096 | 3.3089 | 0.0280 |  118.0x
```

**Comment by Balandat** [comment]
Very nice! Looks like your gpytorch patch caused some unforeseen behavior with our High-Order GP model implementation in BoTorch, which is why the unit tests are broken. I'll take a stab at figuring out what's going on there, hopefully this is a small fix after which we can merge this PR in.

**Comment by valtron** [comment]
Thanks! I also looked into it a little bit. It seems that it failed before; here is a small testcase:

```python
import numpy as np
import torch
import gpytorch
import botorch
print('torch', torch.__version__)
print('gpytorch', gpytorch.__version__)
print('botorch', botorch.__version__)

from botorch.models import HigherOrderGP
from botorch.models.transforms.input import Normalize
from botorch.models.higher_order_gp import FlattenedStandardize

def main():
	n_eval = 10
	
	for d in [1, 2, 3]:
		for k in [1, 2, 3]:
			rng = np.random.default_rng(0)
			train_X = torch.tensor(rng.uniform(size = (5, d)))
			train_Y = torch.tensor(rng.uniform(size = (5, k)))
			
			gp = HigherOrderGP(
				train_X=train_X,
				train_Y=train_Y,
				input_transform=Normalize(d=3),
				# doesn't make a difference
				#outcome_transform=FlattenedStandardize(train_Y.shape[1:]),
			)
			X_eval = torch.tensor(rng.uniform(size = (n_eval, d)))
			
			for mode in ['orig', 'patch']:
				try:
					if mode == 'patch':
						with monkey_patches():
							_ = gp.posterior(X_eval)
					else:
						_ = gp.posterior(X_eval)
				except:
					result = 'fail'
				else:
					result = 'pass'
				
				print("d={} k={} {:<5}: {}".format(d, k, mode, result))
```

Output:
```
torch 1.9.0+cu111
gpytorch 1.5.1
botorch 0.5.1
d=1 k=1 orig : fail
d=1 k=1 patch: fail
d=1 k=2 orig : fail
d=1 k=2 patch: fail
d=1 k=3 orig : fail
d=1 k=3 patch: fail
d=2 k=1 orig : fail
d=2 k=1 patch: fail
d=2 k=2 orig : fail
d=2 k=2 patch: fail
d=2 k=3 orig : fail
d=2 k=3 patch: fail
d=3 k=1 orig : pass
d=3 k=1 patch: pass
d=3 k=2 orig : pass
d=3 k=2 patch: pass
d=3 k=3 orig : pass
d=3 k=3 patch: pass
```

**Comment by valtron** [comment]
The issue might be in `LazyTensor.__getitem__`: line [2110](https://github.com/cornellius-gp/gpytorch/blob/c58dc3a77c7be7ee4434484e359a0bbc1c2d27f6/gpytorch/lazy/lazy_tensor.py#L2110) (and maybe 2106-2107, but those aren't being hit) seem to assume that no new dimensions are introduced, when padding `ndimension - ...`. `index` is `(None, slice())`, which should get turned into `batch_indices = (None,), row_index = slice(), col_index = slice()`, but instead it's `batch_indices = (), row_index = None, col_index = slice()`.

It's possible to work around this by [changing this](https://github.com/cornellius-gp/gpytorch/commit/c58dc3a77c7be7ee4434484e359a0bbc1c2d27f6) so that line becomes a noop, then `TestHigherOrderGP` tests pass:
```diff
     def _unsqueeze_batch(self, dim):
-        return self[(slice(None),) * dim + (None,)]
+        return self[(slice(None),) * dim + (None, slice(None), slice(None))]
```

But regardless, I'm trying to figure out how to change `LazyTensor.__getitem__` to support `None` indexes.

**Comment by Balandat** [comment]
@valtron looks like there is one unused import, other than that this lgtm!

**Comment by valtron** [comment]
Done!
